src/aspire/reconstruction/mean.py

```diff
@@ -1,17 +1,262 @@
+        """
+        :return: r x r matrix of volumes, shaped (r, r, 2L, 2L, 2L).
+        """
+
+        _2L = 2 * self.src.L
+        # Note, because we're iteratively summing it is critical we zero this array.
+        kernel = np.zeros((self.r, self.r, _2L, _2L, _2L), dtype=self.dtype)
+        sq_filters_f = np.square(evaluate_src_filters_on_grid(self.src))
+
+        for k in range(self.r):
+            for j in range(k + 1):
+                for i in range(0, self.src.n, self.batch_size):
+                    _range = np.arange(
+                        i, min(self.src.n, i + self.batch_size), dtype=int
+                    )
+                    pts_rot = rotated_grids(
+                        self.src.L, self.src.rotations[_range, :, :]
+                    )
+                    weights = sq_filters_f[:, :, _range]
+                    weights *= self.src.amplitudes[_range] ** 2
+
+                    if self.src.L % 2 == 0:
+                        weights[0, :, :] = 0
+                        weights[:, 0, :] = 0
+
+                    weights *= (
+                        self.weights[_range, j] * self.weights[_range, k]
+                    ).reshape(1, 1, len(_range))
+
+                    pts_rot = pts_rot.reshape((3, -1))
+                    weights = np.transpose(weights, (2, 0, 1)).flatten()
+
+                    batch_kernel = (
+                        1
+                        / (2 * self.src.L**4)
+                        * anufft(weights, pts_rot[::-1], (_2L, _2L, _2L), real=True)
+                    )
+                    kernel[k, j] += batch_kernel
+
+                    # r x r symmetric
+                    # accumulate batch entries of kernel[k,j] to kernel[j,k]
+                    if j != k:
+                        kernel[j, k] += batch_kernel
+
+        kermat_f = np.zeros((self.r, self.r, _2L, _2L, _2L))
+        logger.info("Computing non-centered Fourier Transform Kernel Mat")
+        for k in range(self.r):
+            for j in range(self.r):
+                # Ensure symmetric kernel
+                kernel[k, j, 0, :, :] = 0
+                kernel[k, j, :, 0, :] = 0
+                kernel[k, j, :, :, 0] = 0
+
+                kernel[k, j] = fft.mdim_ifftshift(kernel[k, j], range(0, 3))
+                kernel_f = fft.fftn(kernel[k, j], axes=(0, 1, 2))
+
+                kernel_f = np.real(kernel_f)
+                kermat_f[k, j] = kernel_f
+
+        return FourierKernelMat(kermat_f, centered=False)
+
+    def src_backward(self):
+        """
+        Apply adjoint mapping to source
+
+        :return: The adjoint mapping applied to the images, averaged over the whole dataset and expressed
+            as coefficients of `basis`.
+        """
+
+        # src_vols_wt_backward
+        mean_b = Volume(
+            np.zeros((self.r, self.src.L, self.src.L, self.src.L), dtype=self.dtype)
+        )
+
+        for k in range(self.r):
+            for i in range(0, self.src.n, self.batch_size):
```

Comment:
Again, consider swapping order to increase image reuse (esp. if `src.images` are generated on the fly).

Reply:
Yep, that is better.

Reply:
Cool. But again probably  want to move the first line out of the inner loop.

Reply:
This has been consolidated into one method, and in that method these lines have been lifted to the batch (i) loop.